Schema spec revision (#400)
xregistry/spec: schema/spec.md

distributed systems. Common schema formats include JSON Schema, JSON
Structure, Apache Avro Schema, Google Protobuf Schema, and XML Schema.

### 1.1. Schemas

Schema registries are generally used to share such schemas amongst multiple
parties.

When schemas are used to drive the serialization and encoding of data, like in
the cases of Apache Avro or Google Protobuf, the deserialization of structured
data from its encoded form requires the schema to be available to the
deserializer. The compactness of these serialization formats is achieved by
externalizing type information into a schema document. The registry allows for a
publisher of data to publish the schema document and pass a reference to it, and
for a consumer of data to retrieve the schema document and use it to decode the
data.

Formats like XML and JSON do not require a schema to decode the data, but
schemas are still very useful to establish a common understanding of the data
structures that are exchanged, provide a foundation for code generation, allow
for validation of the data, and provide an anchor for documentation and semantic
information, like scientific units for numeric values, that goes beyond simple

> labels and data types. Generally, it is a best practice for data structures that
> are exchanged in a distributed systems ought to be described by a schema, even
> if the data serialization model does not require one.

Reviewer: ```suggestion
labels and data types. Generally, it is a best practice for data structures that
are exchanged in a distributed system to be described by a schema, even
if the data serialization model does not require one.
```

### 1.2. Schema References

In the [CloudEvents][CloudEvents dataschema] specification, the `dataschema`

> attribute holds a URI and is specifically thought to reference a schema document

Reviewer: ```suggestion
attribute holds a URI and is specifically meant to reference a schema document
```

residing in a registry. For example, a CloudEvent with a `dataschema` attribute
pointing to a schema version in a schema registry might look like this, using
the schema version's [`self`][xRegistry self] URL as the value of the `dataschema` attribute:

```json
{
    "specversion": "1.0",
    "id": "1234-5678-9012",
    "type": "com.example.event",
    "source": "https://example.com/source",
    "dataschema": "https://example.com/registry/schemagroups/com.example.schemas/schemas/com.example.event/versions/1.0",
    "datacontenttype": "application/vnd.google.protobuf",
    "data_base64": "...base64-encoded-data..."
}
```

Since this URL might be a bit long, the xRegistry core specification allows for
an implementation to provide an alternative, shorter, self-referencing URL that
points to the same schema version, via the [`shortself`][xRegistry shortself]
attribute. The specification is not prescriptive about the format of the shorter

> URL, but it might follow common URL-shorneter practices. With that, the above

Reviewer: ```suggestion
URL, but it might follow common URL-shortening practices. With that, the above
```

    "datacontenttype": "application/vnd.google.protobuf",
    "data_base64": "...base64-encoded-data..."
}
```

### 1.3. Versioning

When schemas are used in a system, they typically evolve over time. Data
structures are extended or modified, with parts added or deprecated or even
removed. Some of these changes are compatible with existing data, while others
are not.

Serialization generally occurs based on a specific schema version that the data
publisher uses. Multiple versions of publishers may exist in the same system,
using different schema versions, which is a common occurrence in systems that
perform live updates. Once data has been published, data serialized based on
several different versions may exist in a system, in queues, in databases, or in
files.

The schema registry therefore allows managing multiple versions of schemas,
declare their lineage, and state the compatibility policy. The compatibility
policy is used to determine whether a schema change is compatible with existing
data, and MAY be enforced by implementations of the schema registry. For this,
this specification leans on the [xRegistry Core][xRegistry Core] specification

> that already defines these versioning mechanisms for any kind of resource.

Reviewer: ```suggestion
that already defines these versioning and compatibility mechanisms for any kind of resource.
```

### 1.4. Document Store

The schema registry is a document store and therefore also has the
[`hasdocument`][xRegistry hasdocument] attribute defined in the xRegistry Core
attribute (implicitly) defined as `true` for the `schema` Resource.

> What this means is that the schema registry immediately yields a document with

Reviewer: ```suggestion
This means that the schema registry immediately yields a document with
```

externalizing type information into a schema document. The registry allows for a
publisher of data to publish the schema document and pass a reference to it, and
for a consumer of data to retrieve the schema document and use it to decode the
data.

Formats like XML and JSON do not require a schema to decode the data, but
schemas are still very useful to establish a common understanding of the data
structures that are exchanged, provide a foundation for code generation, allow
for validation of the data, and provide an anchor for documentation and semantic
information, like scientific units for numeric values, that goes beyond simple
labels and data types. Generally, it is a best practice for data structures that
are exchanged in a distributed system to be described by a schema, even
if the data serialization model does not require one.

### 1.2. Schema References

In the [CloudEvents][CloudEvents dataschema] specification, the `dataschema`
attribute holds a URI and is specifically meant to reference a schema document
residing in a registry. For example, a CloudEvent with a `dataschema` attribute
pointing to a schema version in a schema registry might look like this, using

> the schema version's [`self`][xRegistry self] URL as the value of the `dataschema` attribute:

Reviewer 2: can you have your editor wrap lines like this at 80?

points to the same schema version, via the [`shortself`][xRegistry shortself]
attribute. The specification is not prescriptive about the format of the shorter
URL, but it might follow common URL-shortening practices. With that, the above
example might look like this:

```json
{
    "specversion": "1.0",
    "id": "1234-5678-9012",
    "type": "com.example.event",
    "source": "https://example.com/source",
    "dataschema": "https://example.com/$267shU79S",
    "datacontenttype": "application/vnd.google.protobuf",
    "data_base64": "...base64-encoded-data..."
}
```

### 1.3. Versioning

When schemas are used in a system, they typically evolve over time. Data
structures are extended or modified, with parts added or deprecated or even
removed. Some of these changes are compatible with existing data, while others
are not.

Serialization generally occurs based on a specific schema version that the data
publisher uses. Multiple versions of publishers MAY exist in the same system,
using different schema versions, which is a common occurrence in systems that
perform live updates. Once data has been published, data serialized based on

> several different versions MAY exist in a system, in queues, in databases, or in

Reviewer 2: ```suggestion
several different versions might exist in a system, in queues, in databases, or in
```

declare their lineage, and state the compatibility policy. The compatibility
policy is used to determine whether a schema change is compatible with existing
data, and MAY be enforced by implementations of the schema registry. For this,
this specification leans on the [xRegistry Core][xRegistry Core] specification
that already defines these versioning and compatibility mechanisms for any kind of resource.

### 1.4. Document Store

The schema registry is a document store and therefore also has the
[`hasdocument`][xRegistry hasdocument] attribute defined in the xRegistry Core
attribute (implicitly) defined as `true` for the `schema` Resource.

This means that the schema registry immediately yields a document with
the stored content-type when a client issues a GET request to the
[`self`][xRegistry self] URL of a schema Version. The associated metadata is
returned in the HTTP headers. The [default version][xRegistry default-version]
of the schema Version is returned when the client issues a GET request to the
[`self`][xRegistry self] URL of the `schema` Resource.

This allows to provide external parties with a link that they can use without
needing to know any details about xRegistry.

> Storing a new schema is similarly straightforward for clients that do not know

Reviewer 2: ```suggestion
Storing a new Version of a schema is similarly straightforward for clients that do not know
```

  - [Abstract](#abstract)
  - [Table of Contents](#table-of-contents)
  - [1. Overview](#1-overview)
    - [1.1. Schemas](#11-schemas)
    - [1.2. Schema References](#12-schema-references)
    - [1.3. Versioning](#13-versioning)
    - [1.4. Document Store](#14-document-store)
  - [2. Notations and Terminology](#2-notations-and-terminology)
    - [2.1. Notational Conventions](#21-notational-conventions)
    - [2.2. Terminology](#22-terminology)
      - [2.2.1. Schema](#221-schema)
    - [2.3. Schema Group](#23-schema-group)
  - [3. Schema Registry Model](#3-schema-registry-model)
  - [4. Schema Registry](#4-schema-registry)
    - [4.1. Schema Groups](#41-schema-groups)
    - [4.2. Schema Resources](#42-schema-resources)
      - [4.2.1. `validation`](#421-validation)
      - [4.2.2. `format`](#422-format)
    - [4.3. Schema Formats](#43-schema-formats)
      - [4.3.1. JSON Schema](#431-json-schema)
      - [4.3.2. XML Schema](#432-xml-schema)
      - [4.3.3. Apache Avro Schema](#433-apache-avro-schema)
      - [4.3.4. Protobuf Schema](#434-protobuf-schema)
    - [5. Security Considerations](#5-security-considerations)

## 1. Overview

> A schema registry provides a respository for managing serialization and

Reviewer 2: ```suggestion
A schema registry provides a respository for managing serialization,
```

validation and data type definitions schemas as they are commonly used in
distributed systems. Common schema formats include JSON Schema, JSON

> Structure, Apache Avro Schema, Google Protobuf Schema, and XML Schema.

Reviewer 2: I think we need to make it clear we support more than just the "common" ones

points to the same schema version, via the [`shortself`][xRegistry shortself]
attribute. The specification is not prescriptive about the format of the shorter
URL, but it might follow common URL-shortening practices. With that, the above
example might look like this:

```json
{
    "specversion": "1.0",
    "id": "1234-5678-9012",
    "type": "com.example.event",
    "source": "https://example.com/source",
    "dataschema": "https://example.com/$267shU79S",
    "datacontenttype": "application/vnd.google.protobuf",
    "data_base64": "...base64-encoded-data..."
}
```

### 1.3. Versioning

When schemas are used in a system, they typically evolve over time. Data
structures are extended or modified, with parts added or deprecated or even
removed. Some of these changes are compatible with existing data, while others
are not.

Serialization generally occurs based on a specific schema version that the data
publisher uses. Multiple versions of publishers may exist in the same system,
using different schema versions, which is a common occurrence in systems that
perform live updates. Once data has been published, data serialized based on

> several different versions may exist in a system, in queues, in databases, or in

Reviewer 2: ```suggestion
several different versions might exist in a system, in queues, in databases, or in
```

files.

The schema registry therefore allows managing multiple versions of schemas,
declare their lineage, and state the compatibility policy. The compatibility
policy is used to determine whether a schema change is compatible with existing
data, and MAY be enforced by implementations of the schema registry. For this,
this specification leans on the [xRegistry Core][xRegistry Core] specification
that already defines these versioning and compatibility mechanisms for any kind of resource.

### 1.4. Document Store

The schema registry is a document store and therefore also has the
[`hasdocument`][xRegistry hasdocument] attribute defined in the xRegistry Core
attribute (implicitly) defined as `true` for the `schema` Resource.

This means that the schema registry immediately yields a document with
the stored content-type when a client issues a GET request to the
[`self`][xRegistry self] URL of a schema Version. The associated metadata is
returned in the HTTP headers. The [default version][xRegistry default-version]
of the schema Version is returned when the client issues a GET request to the
[`self`][xRegistry self] URL of the `schema` Resource.

> This allows to provide external parties with a link that they can use without

Reviewer 2: ```suggestion
This enables the ability to provide external parties with a link that they can use without
```

needing to know any details about xRegistry.

Storing a new schema is similarly straightforward for clients that do not know
xRegistry specifics, by simply using a POST against [`self`][xRegistry self] URL
of the `schema` Resource in the simplest case.

> To access the metadata of the `schema` or the schema version as a JSON document,

Reviewer 2: ```suggestion
To access the metadata of the `schema`, or the schema version as a JSON document,
```

and `+` means the preceding attribute MUST appear at least once. The presence
of the `#` character means the remaining portion of the line is a comment.
Whitespace characters in the JSON snippets are used for readability and are
not normative.

### 2.2. Terminology

This specification defines the following terms:

#### 2.2.1. Schema

We use the term **schema** (or schema Resource) in this specification as a
logical grouping of **schema Versions**. A **schema Version** is a concrete
document. The **schema** Resource is a semantic umbrella formed around one or
more concrete schema Version documents that represent iterations of the same logical schema.
Per the definition of the
[`compatibility`][xRegistry compatibility] attribute, all Versions of a single
**schema** MUST adhere to the rules defined by the `compatibility` attribute.
Any breaking change MUST result in a new **schema** Resource being created.

In terms of versioning, you can think of a **schema** as a collection of
versions that are compatible according to the selected `compatibility` mode.
MUST be created, to indicate the breaking change. The [`deprecated`](xRegistry deprecated)

> attribute may be used to indicate the appropriate new schema to use following a breaking change.

Reviewer 2: ```suggestion
attribute MAY be used to indicate the appropriate new schema to use following a breaking change.
```

> MUST be created, to indicate the breaking change.

Reviewer 2: Left over from a previous edit/version, I think.

}
```

## 4. Schema Registry

The Schema Registry is a metadata store for organizing schemas and schema
Versions of any kind; it is a document store.

Implementations of this specification MAY include additional extension
attributes, including the `*` attribute of type `any`.

Since the Schema Registry is an application of the xRegistry specification, all
attributes for Groups, Resources, and Resource Version objects are inherited
from there.

### 4.1. Schema Groups

> The Group (`<GROUP>`) name for the Schema Registry is `schemagroups`. The

Reviewer 2: this is switched up: GROUP->schemagroup, GROUPS->schemagroups.
You may want to explicitly mention one "singular" vs "plural", not just the plural name itself

}
```

## 4. Schema Registry

The Schema Registry is a metadata store for organizing schemas and schema
Versions of any kind; it is a document store.

Implementations of this specification MAY include additional extension
attributes, including the `*` attribute of type `any`.

Since the Schema Registry is an application of the xRegistry specification, all
attributes for Groups, Resources, and Resource Version objects are inherited
from there.

### 4.1. Schema Groups

The Group (`<GROUP>`) name for the Schema Registry is `schemagroups`. The
Schema Group does not have any specific extension attributes.

A schema group is a collection of schemas that are related to each other in
some application-defined way. A Schema Group does not impose any restrictions
on the contained schemas, meaning that a Schema Group MAY contain schemas of
different formats.

> Every schema MUST reside inside a Schema Group.

Reviewer 2: It might be me, but trying to read this as a newbie someone may think that a schema can be part of a Group directly. So, perhaps a tiny bit of clarity:
`Every schema (i.e. schema Resource) MUST reside inside a Schema Group`

`"com.example.event.2024-02"`, so that incompatible, but historically related
schemas can be more easily identified by users and developers. The schema
`versionid` then functions as the semantic minor version identifier.

> which is a `versionid` of the Version that this Version is based on. The `ancestor`

Reviewer 2: something is missing before this line.

attribute permits multiple Versions to reference the same ancestor, and allows for

> implementations to determine the Version's ancestor. See the

Reviewer 2: s/See/Since/ ??

  - MUST be a Resource level attribute defined within the `metaattributes`
    section of the model.

#### 4.2.2. `format`

- Type: String
- Description: Identifies the schema format. In absence of formal media-type
  definitions for several important schema formats, we define a convention here
  to reference schema formats by name and version as `<NAME>/<VERSION>`. This

>   specification defines a set of common [schema format names](#43-schema-formats)

Reviewer 2: While being able to point users to a section # is handy, one downside to having numbers in section titles is that if we ever renumber things ALL docs (especially ones we don't control) will now have bad links - even if the section in question was just moved. Something to think about.

`#TelemetryEvent` references the record declaration of the `TelemetryEvent`
record.
- If the Avro schema document is a local Schema Registry reference like
`#/schemagroups/com.example.telemetry/schemas/com.example.telemetrydata`, in the

> which the reference is already in the form of a URI fragment, the suffix is

Reviewer 3: ```suggestion
`#/schemagroups/com.example.telemetry/schemas/com.example.telemetrydata`, in
which the reference is already in the form of a URI fragment, the suffix is
```

  is appended separated with a colon, for instance
  `.../com.example.telemetrydata:TelemetryEvent`.

Like the [xRegistry Core][xRegistry Core] specification, this specification does not

Like [xRegistry Core][xRegistry Core] specification, this specification does not
explicitly address authentication or authorization levels of users, nor how to
securely protect the APIs.

It is expected that any implementation of this specification will use
authentication and authorization mechanisms that are appropriate for the

> application domain and the deployment environment. This may include, but is not

Reviewer 2: ```suggestion
application domain and the deployment environment. This MAY include, but is not
```
